### DISCUSSION TITLE
About timed events and note events

### DISCUSSION THREAD

**Discussion by Jenix-Park** [melanchall/drywetmidi#186] [Q&A] [answered]
Apologies again for keeping you with my bothers.

One example code from the official page a little confuses me.
As far as I know, timed events contain all the events including note ones.
But why did the following code have to include `ObjectType.Note`?

```
                    using (var manager = new TimedObjectsManager(trackChunk.Events, ObjectType.Note | ObjectType.TimedEvent))
                    {
                        foreach (var obj in manager.Objects)
                        {
                            if (obj is Note note)
                                note.Length -= 10;
                            else if (obj is TimedEvent timedEvent && timedEvent.Event is BaseTextEvent textEvent)
                                textEvent.Text = "Hello";
                        }
                    }
```

Couldn't we just call either `GetTimedEvents()` or `GetObjects(ObjectType.TimedEvent)` and then check if it's the note event via `obj.Event is NoteOnEvent`?
If you say it's because dealing with the Note interface is much easier, then I understand.
But I'm asking this because it seems we would have duplicate note events by doing this way.
Am I wrong about this?

**Comment by melanchall** [comment, marked as answer]
**Note** is not `NoteOnEvent`. Note is a pair of events: Note On and corresponding Note Off one. So `GetTimedEvents()` will return events including `NoteOnEvent` and `NoteOffEvent`, but **event is not a note**. DryWetMIDI provides following key objects:

* `TimedEvent` – just a MIDI event along with absolute time;
* `Note` – pair of timed events holding `NoteOnEvent` and `NoteOffEvent`;
* `Chord` – set of `Notes`s.

So `ObjectType.TimedEvent` will give you just all timed events, but `ObjectType.Note | ObjectType.TimedEvent` will return notes for pairs of linked events and timed events for all other events.

**Reply by Jenix-Park** [reply]
Right after writing this question, I realized how different they are from each other and how helpful `Note` objects are.
But I'm not sure if I understood your last sentence correctly.

If I pass `ObjectType.Note | ObjectType.TimedEvent`, then will the NoteOnEvent/NoteOffEvents be included as `TimedEvents` only when those aren't used to create `Note` objects? So no redundant data between Notes and TimedEvents?

**Reply by melanchall** [reply]
You're absolutely right :)

Simple example:

```
a b A c B d e A f g B h i
```

where upper-case letters are Note On (`A`) and Note Off (`B`) events, and lower-case letters are other MIDI events.

If we use `ObjectType.TimedEvent` option, we'll get just all those events:

```
a b A c B d e A f g B h i
```

But if we use `ObjectType.Note | ObjectType.TimedEvent`, we'll get following objects:

```
a b N1 c d e N2 f g h i
```

where `N1` and `N2` – notes build on `A`/`B` events pairs.

Please see detailed info in the [GetObjects section of the Getting objects](https://melanchall.github.io/drywetmidi/articles/high-level-managing/Getting-objects.html#getobjects) article.

**Reply by Jenix-Park** [reply]
Yeah, I'm reading all the articles.
The documentations are as great as the "DryWetMIDI" plugin itself.
I kept asking questions while I was reading them because I wanted to fully understand and remove any misunderstandings in order to get the most out of DryWetMIDI.
Thanks for this great work!
